src/dawn/engine_dawn.cc

```diff
@@ -14,15 +14,162 @@
+  descriptor.dimension = ::dawn::TextureDimension::e2D;
+  descriptor.size.width = kFramebufferWidth;
+  descriptor.size.height = kFramebufferHeight;
+  descriptor.size.depth = 1;
+  descriptor.arrayLayer = 1;
+  descriptor.format = format;
+  descriptor.levelCount = 1;
+  descriptor.usage = ::dawn::TextureUsageBit::TransferSrc |
+                     ::dawn::TextureUsageBit::OutputAttachment;
+  // TODO(dneto): Get a better message by using the Dawn error callback.
+  *result_ptr = device.CreateTexture(&descriptor);
+  if (*result_ptr)
+    return {};
+  return Result("Dawn: Failed to allocate a framebuffer texture");
+}
+
+// Creates a host-side buffer for the framebuffer, and returns it through
+// |result_ptr|. The buffer will be used as a transfer destination and
+// for mapping-for-read.  Returns a result code.
+Result MakeFramebufferBuffer(const ::dawn::Device& device,
+                             ::dawn::TextureFormat format,
+                             ::dawn::Buffer* result_ptr,
+                             uint32_t* texel_stride_ptr,
+                             uint32_t* row_stride_ptr,
+                             uint32_t* size_ptr) {
+  assert(device);
+  assert(result_ptr);
+  assert(texel_stride_ptr);
+  assert(row_stride_ptr);
+  assert(size_ptr);
+
+  ::dawn::BufferDescriptor descriptor;
+  // TODO(dneto): Handle other formats.
+  if (format != ::dawn::TextureFormat::R8G8B8A8Unorm) {
+    return Result("Dawn::MakeFramebufferBuffer: Unhandled framebuffer format");
+  }
+  // Number of bytes for each texel in the default format.
+  const uint32_t default_texel_bytes = 4;
+
+  uint32_t row_stride = default_texel_bytes * kFramebufferWidth;
+  {
+    // Round up the stride to the minimum image row pitch.
+    const uint32_t spillover = row_stride % kImageRowPitch;
+    if (spillover > 0)
+      row_stride += (kImageRowPitch - spillover);
+    assert(0 == (row_stride % kImageRowPitch));
+  }
+
+  descriptor.size = row_stride * kFramebufferHeight;
+  descriptor.usage =
+      ::dawn::BufferUsageBit::TransferDst | ::dawn::BufferUsageBit::MapRead;
+  *result_ptr = device.CreateBuffer(&descriptor);
+  *texel_stride_ptr = default_texel_bytes;
+  *row_stride_ptr = row_stride;
+  *size_ptr = descriptor.size;
+  return {};
+}
+
+// Result status object and data pointer resulting from a buffer mapping.
+struct MapResult {
+  Result result;
+  const void* data = nullptr;
+};
+
+// Handles the update from an asynchronous buffer map request, updating the
+// state of the MapResult object hidden inside the |userdata| parameter.
+// On a successful mapping outcome, set the data pointer in the map result.
+// Otherwise set the map result object to an error, and the data member is
+// not changed.
+void HandleBufferMapCallback(dawnBufferMapAsyncStatus status,
+                             const void* data,
+                             dawnCallbackUserdata userdata) {
+  MapResult& map_result = *reinterpret_cast<MapResult*>(userdata);
```

Comment:
nit: static_cast userdata to intptr_t before the reinterpret_cast to avoid a warning on 32bit systems. We should really fix this to be a pointer in Dawn. I can't remember why it is a u64